service/dap/server.go

```diff
@@ -2957,6 +3037,54 @@ func (s *Server) doRunCommand(command string, asyncSetupDone chan struct{}) {
 	}
 }
 
+func (s *Server) run(command string, asyncSetupDone chan struct{}) (*api.DebuggerState, error) {
+	// TODO(polina): it appears that debugger.Command doesn't always close
+	// asyncSetupDone (e.g. when having an error next while nexting).
+	// So we should always close it ourselves just in case.
+	defer s.asyncCommandDone(asyncSetupDone)
+	s.setRunning(true)
+	defer s.setRunning(false)
+
+	runningStep := command != api.Continue
+	state, _ := s.debugger.State(true)
+	if state != nil {
+		// If state.NextInProgress, then we are in the middle of a step,
+		// regardless of the command.
+		runningStep = runningStep || state.NextInProgress
+	}
+	// The handleRequest loop will release resumeMu after asyncSetupDone is closed.
+	state, err := s.debugger.Command(&api.DebuggerCommand{Name: command}, asyncSetupDone)
+	for state != nil && state.CurrentThread != nil && err == nil {
+		if bp := state.CurrentThread.Breakpoint; bp != nil && bp.Tracepoint {
+			// If this is a log point, we want to log the message and continue execution.
+			s.handleLogPoint(bp.ID)
```

Comment:
There could be multilpe simultaneous logpoints happening in a single stop, as well as other non-logpoint breakpoints.

Reply:
I was currently planning on not dealing with simultaneous breakpoints since this is already a known limitation.

It definitely does seem worthwhile to handle log points on all threads and make sure that another real breakpoint has not been hit. Will make that change.

Reply:
Done. Added a test as well.